### ISSUE TITLE
Slice before F.scaled_dot_product_attention() to improve the performance

### ISSUE THREAD

**Issue by mzchtx** [Lightning-AI/lit-llama#383]
I think we can slice k, v and mask before calling `F.scaled_dot_product_attention()` to reduce the calculation, otherwise the calculation is the same as max_seq_len even when input_pos is relatively small

https://github.com/Lightning-AI/lit-llama/blob/main/lit_llama/model.py#L209-L235

![image](https://github.com/Lightning-AI/lit-llama/assets/37564754/f5db0a3d-b5df-4f89-97e7-e56415bd6d55)

**Comment by carmocca** [comment]
@mzchtx Would you like to open a PR with your suggested changes?

**Comment by mzchtx** [comment]
> @mzchtx Would you like to open a PR with your suggested changes?

I tested more data and had a more detailed comparison:

1. The current implementation is more efficient When the batch size is relatively small (batch size < 16 or 32). It is IO-bound, so performing more calculations in exchange for IO operations would be more efficient.
2. The IO bottleneck diminishes significantly When the batch size is relatively large (batch > 32). It is advisable to avoid excessive redundant calculations, making slicing a more efficient approach.

Of course, implementing a custom kernel to avoid memory copy generated by slicing and eliminating redundant computations would be the most efficient method. However, it would require additional work. Therefore, we can retain the current implementation.

Below are the detailed data; all experiments were done on A30 with LLaMA 7B.

<img width="833" alt="image" src="https://github.com/Lightning-AI/lit-llama/assets/37564754/d55c03ab-c831-42e5-8b78-110485292499">

<img width="824" alt="image" src="https://github.com/Lightning-AI/lit-llama/assets/37564754/f3fddc12-59a7-48be-8a51-2f579f69988d">

The modified code is shown below:

<img width="732" alt="image" src="https://github.com/Lightning-AI/lit-llama/assets/37564754/4f0c18e1-6b74-4bc8-a806-a49042cfb36a">

**Comment by carmocca** [comment]
@mzchtx The code would need to be indented under the `if` as before. since this is only relevant for the kv-cache case.

Leaving #382 aside, I believe the code should be

```diff
-                input_pos = torch.tensor(max_seq_length - 1, device=input_pos.device)
+                input_pos = torch.tensor([max_seq_length - 1], device=input_pos.device)
                 # shift 1 position to the left
                 cache_k = torch.roll(cache_k, -1, dims=2)
                 cache_v = torch.roll(cache_v, -1, dims=2)
             k = cache_k.index_copy(2, input_pos, k)
             v = cache_v.index_copy(2, input_pos, v)
             kv_cache = k, v
+            input_pos = torch.arange(0, input_pos[-1] + 1, device=input_pos.device)
+            k = k.index_select(2, input_pos)
+            v = v.index_select(2, input_pos)
+            mask = mask.index_select(3, input_pos)
```

What is OPT and Baseline exactly in your tables?

**Comment by carmocca** [comment]
From playing with this, the generated outputs are not the same, meaning that this is not numerically equivalent. However, it's hard to tell if they are worse or just different.

Lastly, you mention that the benefit is mainly for large batch sizes. Our generation scripts do not support batched inference, and the kv-cache is only used during inference.

I am hesitant to make this change

**Comment by carmocca** [comment]
I stumbled upon this issue: https://github.com/pytorch/pytorch/issues/103082, it might explain the numerical difference.

**Comment by mzchtx** [comment]
> From playing with this, the generated outputs are not the same, meaning that this is not numerically equivalent. However, it's hard to tell if they are worse or just different.
> 
> Lastly, you mention that the benefit is mainly for large batch sizes. Our generation scripts do not support batched inference, and the kv-cache is only used during inference.
> 
> I am hesitant to make this change

Yes, this does only have an impact on Inference, and only on large batch sizes, whereas in real scenarios batch sizes are generally smaller, so I agree with your idea and can leave the status quo.
